R/convergence.R

```diff
@@ -316,6 +316,144 @@ quantile2 <- function(x, probs = c(0.05, 0.95), names = TRUE, ...) {
+#'
+#' @param nsim number of R* values in returned vector if uncertainty is true.
+#'
+#' @param training_proportion proportion of iterations used to train classifier.
+#'
+#' @details `rstar` provides a measure of MCMC convergence based on whether it is
+#' possible to determine the chain that generated a particular draw with a probability
+#' greater than chance. To do so, it fits a machine learning classifier to a subset of
+#' the original MCMC draws (the training set) and evaluates its predictive accuracy on
+#' the remaining draws (the testing set). If predictive accuracy exceeds chance (i.e.
+#' predicting the chain that generated a draw uniformly at random), the diagnostic
+#' measure, R* > 1, indicating that convergence has yet to occur.
+#'
+#' The statistic, R*, is stochastic, meaning that each time the test is run, unless
+#' the random seed is fixed, it will generally produce a different result. To minimise
+#' the implications of this stochasticity, it is recommended to repeatedly run this
+#' function to calculate a distribution of R*; alternatively, an approximation to this
+#' distribution can be obtained by setting uncertainty = T.
+#'
+#' By default, a random forest classifier is used, which tends to perform best for
+#' target distribution dimensionalities of around 4 and above. For lower dimensional
+#' targets, gradient boosted models (called via `method=\"gbm\"`) tend to have a
+#' higher classification accuracy. On a given MCMC sample, it is recommended to
+#' try both of these classifiers.
+#'
+#' @return A single numeric R* value (by default) or a vector of values (if
+#' uncertainty is true).
+#'
+#' @references Ben Lambert, Aki Vehtari (2020) R*: A robust MCMC convergence
+#' diagnostic with uncertainty using gradient-boosted machines
+#' \emph{arXiv preprint} \code{arXiv:TBD}
+#'
+#' @examples
+#' x <- example_draws("eight_schools")
+#' rstar(x)
+#' rstar(x, split_chains = F)
+#' rstar(x, method = "gbm")
+#' # can pass additional arguments to methods
+#' rstar(x, method = "gbm", verbose = F)
+#'
+#' # with uncertainty, returns a vector of R* values
+#' hist(rstar(x, uncertainty = T))
+#' hist(rstar(x, uncertainty = T, nsim = 100))
+#'
+#' # can use other classification methods in caret library
+#' rstar(x, method = "knn")
+#'
+#' @export
+rstar <- function(x, split_chains=T, uncertainty=F, method=NULL, hyperparameters=NULL,
+                            training_proportion=0.7, nsim=1000, ...){
+
+  if(!requireNamespace("caret", quietly = T))
+    stop("Package \"caret\" needed for this function to work. Please install it.")
+  require(caret)
+
+  nsim <- round(nsim)
+  if(nsim < 1)
+    stop("nsim must exceed 1.")
+
+  if(training_proportion <= 0 || training_proportion >= 1)
+    stop("training_proportion must be greater than zero and less than 1.")
+
+  x <- as_draws_df(x)
+  if(split_chains)
+    x <- split_chains_draws_df(x)
+
+  # if only 1 param, add in a column of random noise
+  nparams <- nvariables(x)
```

Comment:
could we rename the object `nparams` to `nvariables` for increase consistency across the posterior package code?

Reply:
Happy to do so. I didn't want to here because that would then change nvariables from a (posterior) function to a value. Are you happy with that?

Reply:
The scoping of vectors and functions is different in R so that `nvariables(...)` still calls the function even if another non-function object in scope has the same name. Alternatively, we could call the value `nvars`. Both options are fine for me. I just don't like the `nparams` since calling things 'variables' instead of 'parameters' was a design decision in posterior.

Reply:
Cool. Done.